server/src/main/java/org/elasticsearch/node/NodeService.java

```diff
@@ -226,4 +238,27 @@ public boolean awaitClose(long timeout, TimeUnit timeUnit) throws InterruptedExc
         return indicesService.awaitClose(timeout, timeUnit);
     }
 
+    private List<NodeStatsPlugin> getNodeStatsPlugins(final PluginsService pluginsService) {
+        return pluginsService.filterPlugins(NodeStatsPlugin.class).toList();
+    }
+
+    private Optional<PluginNodeStats> getPluginStats() {
+        if (nodeStatsPlugins.isEmpty()) {
+            return Optional.empty();
+        }
+
+        final Map<String, NodeStatsPlugin.Stats> pluginNameToStats = new LinkedHashMap<>();
+        for (final NodeStatsPlugin plugin : nodeStatsPlugins) {
+            final String name = Objects.requireNonNull(plugin.getNodeStatsPluginName());
+            final NodeStatsPlugin.Stats stats = Objects.requireNonNull(plugin.getPluginNodeStats());
+
+            if (pluginNameToStats.containsKey(name)) {
+                throw new IllegalArgumentException(
+                    Strings.format("Duplicate node stats plugin name: %s. Plugins: %s", name, pluginNameToStats.keySet())
+                );
+            }
+            pluginNameToStats.put(name, stats);
+        }
+        return Optional.of(new PluginNodeStats(Collections.unmodifiableMap(pluginNameToStats)));
```

Comment:
We should make the `PluginNodeStats` class the one responsible for making the map immutable, rather than the one constructing it.